frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -2635,67 +2635,87 @@ static const Type* getNumericType(Context* context,
   return nullptr;
 }
 
+/*
+  gets either a c_ptr or c_ptrConst type depending on the name in the CallInfo
+*/
 static const Type* getCPtrType(Context* context,
                                const AstNode* astForErr,
                                const CallInfo& ci) {
   UniqueString name = ci.name();
+  bool isConst;
 
   if (name == USTR("c_ptr")) {
-    // Should we compute the generic version of the type (e.g. c_ptr(?))
-    bool useGenericType = false;
-
-    // There should be 0 or 1 actuals depending on if it is ?
-    if (ci.hasQuestionArg()) {
-      // handle c_ptr(?)
-      if (ci.numActuals() != 0) {
-        context->error(astForErr, "invalid c_ptr type construction");
-        return ErroneousType::get(context);
-      }
-      useGenericType = true;
-    } else {
-      // handle c_ptr(?t) or c_ptr(eltT)
-      if (ci.numActuals() != 1) {
-        context->error(astForErr, "invalid c_ptr type construction");
-        return ErroneousType::get(context);
-      }
+    isConst = false;
+  } else if (name == USTR("c_ptrConst")) {
+    isConst = true;
+  } else {
+    return nullptr;
+  }
+  // Should we compute the generic version of the type (e.g. c_ptr(?)/c_ptrConst(?)
+  bool useGenericType = false;
 
-      QualifiedType qt = ci.actual(0).type();
-      if (qt.type() && qt.type()->isAnyType()) {
-        useGenericType = true;
-      }
+  // There should be 0 or 1 actuals depending on if it is ?
+  if (ci.hasQuestionArg()) {
+    // handle c_ptr(?)/c_ptrConst(?)
+    if (ci.numActuals() != 0) {
+      context->error(astForErr, "invalid %s type construction", name.c_str());
+      return ErroneousType::get(context);
+    }
+    useGenericType = true;
+  } else {
+    // handle c_ptr(?t) or c_ptr(eltT)/c_ptrConst(?t) or c_ptrConst(eltT)
+    if (ci.numActuals() != 1) {
+      context->error(astForErr,"invalid %s type construction", name.c_str());
+      return ErroneousType::get(context);
     }
 
-    if (useGenericType) {
-      return CPtrType::get(context);
+    QualifiedType qt = ci.actual(0).type();
+    if (qt.type() && qt.type()->isAnyType()) {
+      useGenericType = true;
     }
+  }
 
-    QualifiedType qt;
-    if (ci.numActuals() > 0)
-      qt = ci.actual(0).type();
+  if (useGenericType) {
+    return isConst ? CPtrType::getConst(context) : CPtrType::get(context);
+  }
 
-    const Type* t = qt.type();
-    if (t == nullptr) {
-      // Details not yet known so return UnknownType
-      return UnknownType::get(context);
-    }
-    if (t->isUnknownType() || t->isErroneousType()) {
-      // Just propagate the Unknown / Erroneous type
-      // without raising any errors
-      return t;
-    }
+  QualifiedType qt;
+  if (ci.numActuals() > 0)
+    qt = ci.actual(0).type();
```

Comment:
The logic above seems like we wouldn't have made it here if we didn't already have 1 actual. I think so because:

1. if ci has a question arg, we use generic type, and return above.
2. Else, if the num actuals isn't 1, we error, and return.

Thus, seems like it's safe to assume that `ci.numActuals() > 0`.

Reply:
I see, thanks! I might just put an assert before it then, in case we ever mess with the conditionals above and allow something without actuals through.